libgnucash/app-utils/gnc-option-impl.hpp

```diff
@@ -0,0 +1,1066 @@
+                                       GncOptionUIType ui_type = GncOptionUIType::INTERNAL
+        ) :
+        OptionClassifier{section, name, key, doc_string},
+        m_ui_type{ui_type}, m_value{value}, m_default_value{value},
+        m_validator{validator}
+        {
+            if (!this->validate(value))
+            throw std::invalid_argument("Attempt to create GncValidatedOption with bad value.");
+        }
+    GncOptionValidatedValue<ValueType>(const char* section, const char* name,
+                                       const char* key, const char* doc_string,
+                                       ValueType value,
+                                       std::function<bool(ValueType)>validator,
+                                       ValueType val_data) :
+        OptionClassifier{section, name, key, doc_string},
+        m_ui_type{GncOptionUIType::INTERNAL}, m_value{value},
+        m_default_value{value}, m_validator{validator}, m_validation_data{val_data}
+    {
+            if (!this->validate(value))
+            throw std::invalid_argument("Attempt to create GncValidatedOption with bad value.");
+    }
+    GncOptionValidatedValue<ValueType>(const GncOptionValidatedValue<ValueType>&) = default;
+    GncOptionValidatedValue<ValueType>(GncOptionValidatedValue<ValueType>&&) = default;
+    GncOptionValidatedValue<ValueType>& operator=(const GncOptionValidatedValue<ValueType>&) = default;
+    GncOptionValidatedValue<ValueType>& operator=(GncOptionValidatedValue<ValueType>&&) = default;
+    ValueType get_value() const { return m_value; }
+    ValueType get_default_value() const { return m_default_value; }
+    bool validate(ValueType value) const { return m_validator(value); }
+    void set_value(ValueType value)
+    {
+        if (this->validate(value))
+            m_value = value;
+        else
+            throw std::invalid_argument("Validation failed, value not set.");
+    }
+    void set_default_value(ValueType value)
+    {
+        if (this->validate(value))
+            m_value = m_default_value = value;
+        else
+            throw std::invalid_argument("Validation failed, value not set.");
+    }
+    void reset_default_value() { m_value = m_default_value; }
+    bool is_changed() const noexcept { return m_value != m_default_value; }
+    GncOptionUIType get_ui_type() const noexcept { return m_ui_type; }
+    void make_internal() { m_ui_type = GncOptionUIType::INTERNAL; }
+private:
+    GncOptionUIType m_ui_type;
+    ValueType m_value;
+    ValueType m_default_value;
+    std::function<bool(ValueType)> m_validator;                         //11
+    ValueType m_validation_data;
+};
+
+QofInstance* qof_instance_from_string(const std::string& str,
+                                      GncOptionUIType type);
+QofInstance* qof_instance_from_guid(GncGUID*, GncOptionUIType type);
+std::string qof_instance_to_string(const QofInstance* inst);
```

Comment:
Those look like they could move to qof-instance at some point ?

Reply:
Certainly not the first two, GncOptionUIType is an implementation detail that we wouldn't want to leak into the wider world. The last one maybe, if we find a use for it somewhere else.

Reply:
Ah, indeed. That makes sense.